<a href="https://colab.research.google.com/github/layal-z/AAI614_Zeineddine/blob/data_vision/Graded_Assessment2_Facial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Graded Assignment 2: Building a Face Recognition System with DeepFace

In the digital era, face recognition systems are widely used in various applications, including security systems and social media platforms. The ability to accurately identify and authenticate individuals is crucial for enhancing safety and improving user experiences. In this graded assignment, you will build a reliable face verification and recognition system using DeepFace, a powerful library that simplifies facial recognition tasks.

DeepFace leverages several advanced models, including FaceNet and VGG-Face, to generate face embeddings, which are compact and discriminative feature representations of faces. These embeddings are unique to each individual, making it easier to compare and identify faces. In this assignment, you will practice building a face verification system to confirm a person's identity and a face recognition system to identify and label individuals from a dataset.

Main Objectives:

- Build a Face Recognition System: Create a face recognition system capable of identifying a person from a given set of faces. This will include matching the embeddings of an unknown face to a database of known embeddings and identifying the most similar one.

- Compare Models: Compare the performance of different models in face recognition and verification tasks.


- Evaluation and Performance Analysis: Assess your findings to identify the advantages and disadvantages of your implementation.

Let's get started!

To install the DeepFace library, use the command: !pip install deepface.

In [2]:
!pip install deepface

  Using cached deepface-0.0.100-py3-none-any.whl.metadata (33 kB)
Using cached deepface-0.0.100-py3-none-any.whl (170 kB)


The following code defines a function called face_verification, which uses the DeepFace library to verify whether two face images belong to the same person. Here’s a brief breakdown of the code:

- Imports: The function imports the DeepFace class from the DeepFace library, which provides tools for face recognition and verification.

- Function Definition: face_verification(image_path1, image_path2, model_name='VGG-Face', distance_metric='cosine') accepts two image paths as inputs along with optional parameters for the model name and distance metric. By default, it uses the 'VGG-Face' model and the 'cosine' distance metric.

- Verification Process: The function calls DeepFace.verify, passing the two image paths and specified parameters. This function performs the face verification and returns a result dictionary containing details such as the distance between embeddings and whether the two images are verified as the same person.

- Extracting Results: It extracts the distance and verified values from the result. The distance indicates how similar the two faces are based on the embeddings, while verified is a boolean indicating if they are considered the same person.

- Displaying Results: The function checks the verified value:
If True, it prints a success message with the distance and the verification threshold. If False, it prints a failure message with the distance and the threshold.

- Return Value: Finally, the function returns True if the verification is successful and False otherwise as well as calculated distance.

In [3]:
from deepface import DeepFace

def face_verification(image_path1, image_path2, model_name='VGG-Face', distance_metric='cosine'):
    # Use DeepFace's built-in verification and distance metric
    result = DeepFace.verify(image_path1, image_path2, model_name=model_name, distance_metric=distance_metric, enforce_detection=False)

    # Extract the relevant details
    dist = result['distance']
    verified = result['verified']

    # Display results
    if verified:
        print(f"Verification successful. Distance: {dist:.2f} (Threshold: {result['threshold']})")
    else:
        print(f"Verification not successful. Distance: {dist:.2f} (Threshold: {result['threshold']})")

    # Return both verification result and distance
    return verified, dist

## Graded Cell: Building a Database of Individuals [10 points]
Create a database of individuals:

- Collect Images of Individuals: Gather images of **at least 5** individuals whose face embeddings you want to include in the database. Ensure that the images are of good quality and contain only the face of the individual without any obstructions.

- Add Entries to the Database: For each individual, use the individual's name as the key and the image path as value.

- Repeat for Each Individual: Repeat the process for each individual whose identities you want to include in the database.

In [40]:
## Create the database
database = {}

database = {
    #"trump": "/content/trump.jpg",
    # Add more identities and paths as needed
    "Ben Affleck": "/content/BenA_train.jpg",
    "Michelle Obama": "/content/Michelle_Train.jpg",
    "Benson Boone": "/content/BensonB_train.jpg",
    "Taylor Swift": "/content/TaylorS_train.jpg",
    "Leonardo DiCaprio": "/content/LeoDC_train.jpg"
}
print(database)
print("Number of individuals:", len(database))


{'Ben Affleck': '/content/BenA_train.jpg', 'Michelle Obama': '/content/Michelle_Train.jpg', 'Benson Boone': '/content/BensonB_train.jpg', 'Taylor Swift': '/content/TaylorS_train.jpg', 'Leonardo DiCaprio': '/content/LeoDC_train.jpg'}
Number of individuals: 5


## Graded Cell: Face Recognition System [35 points]

The face_recognition function is designed to identify a person in a given image by comparing it with images stored in a database.

Parameters:

- image_path: A string representing the path to the image to be recognized.
- database: A dictionary where the keys are identities (names of individuals) and the values are the paths to their corresponding known images.
- model_name: A string specifying the model to use for generating embeddings (default is 'VGG-Face').
- distance_metric: A string indicating the metric to calculate distance (options include 'cosine', 'euclidean', etc.).

Returns:

A string representing the identity of the recognized person or "Unknown" if no match is found.

The function iterates through the database, using a face verification method to compare the input image with each known image. If a match is found, it returns the identity of the recognized person; otherwise, it returns "Unknown."

In [41]:
def face_recognition(image_path, database, model_name='VGG-Face', distance_metric='cosine'):
    """
    Performs face recognition to identify the person in the given image by comparing it with images in the database.

    Parameters:
    - image_path: str, path to the image to be recognized
    - database: dict, where keys are identities and values are the paths to their known images
    - model_name: str, the model to use for generating embeddings (default is 'VGG-Face')
    - distance_metric: str, the metric to calculate distance ('cosine', 'euclidean', etc.)
    - threshold: float, the threshold for verification (default is 0.68)

    Returns:
    - str: the identity of the recognized person, or "Unknown" if no match is found
    """
    recognized_identity = "Unknown"

    # Loop through the database to verify each identity
    for identity, known_image_path in database.items(): # Use the face_verification function to check each identity
      verified, _ = face_verification(
            image_path,
            known_image_path,
            model_name=model_name,
            distance_metric=distance_metric,
        )
      # If the two images belong to the same person,
        # return that person's identity
      if verified:
        recognized_identity = identity
        break



    return recognized_identity

## Graded Cells: Testing the Face Verification System [5 points]
Evaluate the verification system's performance by testing it on three distinct sample test images. These images should belong to individuals already present in the database but must be different from the ones used in the database.

In [42]:
#Test1
is_match, test1= face_verification(
    "/content/BenA_train.jpg",
    "/content/BenA_test.jpg",
)

if is_match:
    print(f"The provided images are verified as the same identity.")
    print(f"Distance of {test1:.2f} falls within the acceptable threshold of 0.68.")
else:
    print(f"The provided images do not match. Distance of {test1:.2f} exceeds the threshold of 0.68.")

Verification successful. Distance: 0.45 (Threshold: 0.68)
The provided images are verified as the same identity.
Distance of 0.45 falls within the acceptable threshold of 0.68.


In [43]:
# Test 2
is_match, test2=face_verification(
    "/content/Michelle_Test.jpg",
    "/content/BensonB_train.jpg",
)
if is_match:
    print(f"The provided images are verified as the same identity.")
    print(f"Distance of {test2:.2f} falls within the acceptable threshold of 0.68.")
else:
    print(f"The provided images do not match. Distance of {test2:.2f} exceeds the threshold of 0.68.")


Verification not successful. Distance: 0.94 (Threshold: 0.68)
The provided images do not match. Distance of 0.94 exceeds the threshold of 0.68.


In [44]:
# Test 3
is_match, test3= face_verification(
    "/content/BenA_train.jpg",
    "/content/LeoDC_test.jpg",
)
if is_match:
    print(f"The provided images are verified as the same identity.")
    print(f"Distance of {test3:.2f} falls within the acceptable threshold of 0.68.")
else:
    print(f"The provided images do not match. Distance of {test3:.2f} exceeds the threshold of 0.68.")

Verification not successful. Distance: 0.76 (Threshold: 0.68)
The provided images do not match. Distance of 0.76 exceeds the threshold of 0.68.


## Graded Cells: Testing the Face Recognition System [5 points]

Evaluate the recognition system's performance by testing it on three distinct sample test images. These images should belong to individuals already present in the database but must be different from the ones used in the database.

In [45]:
# Test 1
#the face verification prints the outcome of every instance, since Ben Affleck is the first item in the database, we will see one outcome
Test_recognition1 =face_recognition(
    "/content/BenA_test.jpg",
    database,
    model_name="VGG-Face",
    distance_metric="cosine"
)

print("Recognized identity:", Test_recognition1)

Verification successful. Distance: 0.45 (Threshold: 0.68)
Recognized identity: Ben Affleck


In [46]:
# Test 2
# taylor swift is the 4th in the database, so we should see 4 outcomes with one successful
Test_recognition2=face_recognition(
    "/content/TaylorS_test.jpg",
    database,
    model_name="VGG-Face",
    distance_metric="cosine"
)
print("Recognized identity:", Test_recognition2)

Verification not successful. Distance: 0.95 (Threshold: 0.68)
Verification not successful. Distance: 0.95 (Threshold: 0.68)
Verification not successful. Distance: 0.91 (Threshold: 0.68)
Verification successful. Distance: 0.44 (Threshold: 0.68)
Recognized identity: Taylor Swift


In [47]:
# Test 3
#Michelle Obama is the 2nd in the database, so we should see 2 outcomes with one successful
Test_recognition3=face_recognition(
    "/content/Michelle_Test.jpg",
    database,
    model_name="VGG-Face",
    distance_metric="cosine"
)
print("Recognized identity:", Test_recognition3)

Verification not successful. Distance: 0.97 (Threshold: 0.68)
Verification successful. Distance: 0.38 (Threshold: 0.68)
Recognized identity: Michelle Obama


## Graded Cells: Comparison of Different Models [35 points]

Compare the performance of various pre-trained models (e.g., VGG-Face, Facenet, ArcFace) on the 5 images in the database. Evaluate the models based on metrics such as accuracy, average distance, and processing time. You can use either Euclidean or cosine similarity as distance metric.

In [52]:
# The face_verification function contains print statements, which make the
# model-comparison output lengthy. contextlib.redirect_stdout temporarily
# captures these messages, while pandas presents the final metrics in a table.
import time
import numpy as np
import pandas as pd
import contextlib
import io

# Models in comparison
models = ["VGG-Face", "Facenet", "ArcFace"]

# the test images
test_images = {
    "Ben Affleck": "/content/BenA_test.jpg",
    "Michelle Obama": "/content/Michelle_Test.jpg",
    "Benson Boone": "/content/BensonB_test.jpg",
    "Taylor Swift": "/content/TaylorS_test.jpg",
    "Leonardo DiCaprio": "/content/LeoDC_test.jpg"
}

# collecting the results of each model
model_results = {}

for model_name in models:
    print("\nTesting model:", model_name)

    correct_predictions = 0
    distances = []

    start_time = time.time()

    for true_identity, test_image_path in test_images.items():
      # hide the messy print statements by redirecting the standard output
        with contextlib.redirect_stdout(io.StringIO()):
          # Recognize the identity of the test image
          predicted_identity = face_recognition(test_image_path,database,model_name=model_name,distance_metric="cosine")

          # Calculate the distance between the test image
          # and its correct database reference image
          _, distance = face_verification(test_image_path, database[true_identity],model_name=model_name,distance_metric="cosine")

        # Count correct recognition results
        if predicted_identity == true_identity:
            correct_predictions += 1


        distances.append(distance)

    end_time = time.time()

    # Calculate evaluation metrics
    accuracy = correct_predictions *100 / len(test_images)
    average_distance = np.mean(distances)
    processing_time = end_time - start_time

    model_results[model_name] = {
        "Accuracy %": accuracy,
        "Average Distance": average_distance,
        "Processing Time": processing_time
    }
    print("Testing",model_name,"Completed")

results_df = pd.DataFrame(model_results).T

print("FINAL EVALUATION METRICS")
display(results_df)

    #print("\nResults for", model_name)
    #print(f"Accuracy: {accuracy * 100:.2f}%")
    #print(f"Average Distance: {average_distance:.4f}")
    #print(f"Processing Time: {processing_time:.2f} seconds")


Testing model: VGG-Face
Testing VGG-Face Completed

Testing model: Facenet
Testing Facenet Completed

Testing model: ArcFace
Testing ArcFace Completed
FINAL EVALUATION METRICS


,Accuracy %,Average Distance,Processing Time
VGG-Face,100.0,0.386687,38.982432
Facenet,100.0,0.192653,25.910588
ArcFace,100.0,0.418210,25.884269


## Graded Cells: Conclusions [10 points]:

**Conclusions on Model Effectiveness:**

All three models achieved 100% accuracy, showing that each one successfully recognized the five test images in this experiment. Since the dataset was small and contained clear images, accuracy alone is not enough to distinguish between the models. The average matching distance and processing time provide a better comparison.

Facenet produced the lowest average distance (0.193), meaning that, for this dataset, images of the same person were mapped closer together in the embedding space than with VGG-Face (0.387) or ArcFace (0.418). This result is consistent with Facenet's Siamese-network architecture and triplet loss training, which aim to minimize the distance between images of the same identity while maximizing the distance between different identities. Although VGG-Face and ArcFace also classified every image correctly, their embeddings were farther apart on average.

In terms of computational efficiency, Facenet (25.91 s) and ArcFace (25.88 s) performed almost identically and were significantly faster than VGG-Face (38.98 s). VGG-Face is based on an older VGG-style architecture with a larger number of parameters, which generally requires more computation to generate facial embeddings.

It is important to note that the distance values should only be compared within this experiment because each model generates embeddings differently and uses its own verification threshold. Likewise, the perfect accuracy should not be generalized since only five identities with relatively clear images were tested. If the dataset included more people or greater variation in lighting, pose, facial expression, image quality, or partial occlusion, the matching distances would likely increase and the differences between the models might become more apparent.

Overall, Facenet provided the best balance in this experiment. It achieved the same perfect accuracy as the other models while producing the lowest average matching distance and maintaining processing times comparable to ArcFace. However, a larger and more challenging dataset would be required before concluding that one model consistently outperforms the others.

